File: sdk-python/arcflow/_internal/exec_config.py
"""Serialize Sprint 7 execution options for the native binding."""

from __future__ import annotations

import json
from typing import Any

from arcflow.retry import BackoffStrategy, ConstantBackoff, ExponentialBackoff, LinearBackoff


def backoff_to_dict(backoff: BackoffStrategy) -> dict[str, Any]:
    if isinstance(backoff, ExponentialBackoff):
        return {
            "kind": "exponential",
            "base_ms": backoff.base_ms,
            "multiplier": backoff.multiplier,
            "max_ms": backoff.max_ms,
            "jitter_ms": backoff.jitter_ms,
        }
    if isinstance(backoff, LinearBackoff):
        return {
            "kind": "linear",
            "base_ms": backoff.base_ms,
            "increment_ms": backoff.increment_ms,
            "max_ms": backoff.max_ms,
            "jitter_ms": backoff.jitter_ms,
        }
    if isinstance(backoff, ConstantBackoff):
        return {
            "kind": "constant",
            "delay_ms": backoff.delay_ms,
            "jitter_ms": backoff.jitter_ms,
        }
    raise TypeError(f"Unknown backoff type: {type(backoff)!r}")


def build_exec_config_json(
    *,
    retry: tuple[int, BackoffStrategy] | None,
    workflow_timeout_seconds: float | None,
    step_timeout_seconds: float | None,
    recovery_enabled: bool,
    test: dict[str, Any] | None = None,
) -> str | None:
    payload: dict[str, Any] = {}
    if retry is not None:
        max_attempts, backoff = retry
        payload["retry"] = {
            "max_attempts": max_attempts,
            "backoff": backoff_to_dict(backoff),
        }
    if workflow_timeout_seconds is not None:
        payload["workflow_timeout_secs"] = workflow_timeout_seconds
    if step_timeout_seconds is not None:
        payload["step_timeout_secs"] = step_timeout_seconds
    if recovery_enabled:
        payload["recovery_enabled"] = True
    if test is not None:
        payload["test"] = test
    if not payload:
        return None
    return json.dumps(payload)

File: server/arcflow-server/src/exec_config.rs
//! Parse execution config JSON from REST requests.

use std::time::Duration;

use arcflow_core::retry::{BackoffStrategy, RetryConfig};
use arcflow_core::retry::TimeoutConfig as RetryTimeoutConfig;
use arcflow_core::workflow::{ExecutionConfig, StreamConfig};
use serde::Deserialize;

#[derive(Debug, Deserialize)]
struct StreamJson {
    enabled: bool,
}

#[derive(Debug, Deserialize)]
struct ExecConfigJson {
    retry: Option<RetryJson>,
    workflow_timeout_secs: Option<f64>,
    step_timeout_secs: Option<f64>,
    recovery_enabled: Option<bool>,
    stream: Option<StreamJson>,
}

#[derive(Debug, Deserialize)]
struct RetryJson {
    max_attempts: u32,
    backoff: BackoffJson,
}

#[derive(Debug, Deserialize)]
#[serde(tag = "kind", rename_all = "snake_case")]
enum BackoffJson {
    Exponential {
        base_ms: u64,
        multiplier: f64,
        max_ms: u64,
        jitter_ms: u64,
    },
    Linear {
        base_ms: u64,
        increment_ms: u64,
        max_ms: u64,
        jitter_ms: u64,
    },
    Constant {
        delay_ms: u64,
        jitter_ms: u64,
    },
}

impl BackoffJson {
    fn into_strategy(self) -> BackoffStrategy {
        match self {
            BackoffJson::Exponential {
                base_ms,
                multiplier,
                max_ms,
                jitter_ms,
            } => BackoffStrategy::Exponential {
                base_ms,
                multiplier_x100: (multiplier * 100.0) as u64,
                max_ms,
                jitter_ms,
            },
            BackoffJson::Linear {
                base_ms,
                increment_ms,
                max_ms,
                jitter_ms,
            } => BackoffStrategy::Linear {
                base_ms,
                increment_ms,
                max_ms,
                jitter_ms,
            },
            BackoffJson::Constant {
                delay_ms,
                jitter_ms,
            } => BackoffStrategy::Constant { delay_ms, jitter_ms },
        }
    }
}

pub fn parse_exec_config(value: Option<serde_json::Value>) -> Result<ExecutionConfig, String> {
    let Some(json) = value else {
        return Ok(ExecutionConfig::default());
    };
    let parsed: ExecConfigJson =
        serde_json::from_value(json).map_err(|e| format!("Invalid exec_config: {e}"))?;
    let retry = parsed.retry.map(|r| RetryConfig {
        max_attempts: r.max_attempts,
        backoff: r.backoff.into_strategy(),
    });
    let mut timeouts = RetryTimeoutConfig::default();
    if let Some(secs) = parsed.workflow_timeout_secs {
        timeouts.workflow_timeout = Some(Duration::from_secs_f64(secs));
    }
    if let Some(secs) = parsed.step_timeout_secs {
        timeouts.step_timeout = Some(Duration::from_secs_f64(secs));
    }
    Ok(ExecutionConfig {
        retry,
        timeouts,
        recovery_enabled: parsed.recovery_enabled.unwrap_or(false),
        run_id: None,
        test: None,
        stream: parsed.stream.map(|s| StreamConfig {
            enabled: s.enabled,
        }),
    })
}
